add custom resolve option for node/nodes

JipSterk opened the pull request:
A draft for #664 the only thing i do not know is how to implement type inference.
For now i copied one of the tests & examples but i could provide other tests if needed.
Also i added a small bit to the plugins docs but haven't mentioned that a custom resolve for `node/nodes` isn't something most people would want.

hayes commented on packages/plugin-relay/src/schema-builder.ts:
```diff
@@ -156,6 +156,7 @@ schemaBuilderProto.nodeInterfaceRef = function nodeInterfaceRef() {
           id: t.arg.id({ required: true }),
         },
         resolve: async (root, args, context, info) =>
+          this.options.relayOptions?.nodeQueryOptions?.resolve?.(String(args.id), context) ??
```
this will fallback to the default resolver if the custom resolver returns null or undefined

hayes commented on packages/plugin-relay/src/schema-builder.ts:
```diff
@@ -172,9 +173,10 @@ schemaBuilderProto.nodeInterfaceRef = function nodeInterfaceRef() {
       },
       resolve: async (root, args, context, info) =>
-        (await resolveNodes(this, context, info, args.ids as string[])) as Promise<
+        this.options.relayOptions?.nodesQueryOptions?.resolve?.(args.ids as string[], context) ??
```
I think it makes more sense to decode the ids before passing them to the default resolver

hayes replied:
I think maybe the correct thing here would be to define the id(s) args using t.arg.globalID(s), so they are automatically decoded, resolveNodes would need to be updated to expect the ids to be decoded, but that is an easy change.  

The custom resolver should probably just take all the normal resolver args:
If you do that, then I think the better change here would be something like resolve: `this.option.relayOptions?.nodeQueryOptions?.resolve ?? async ...(default resolver here)`.

JipSterk replied:
i have tried to implement the custom resolve like this here: 
```typescript
resolve:
  this.options.relayOptions?.nodesQueryOptions?.resolve ??
  (async (root, args, context, info) =>
    (await resolveNodes(this, context, info, args.ids as string[])) as Promise<
      ObjectParam<SchemaTypes>
    >[]),
```
but it would throw an error about the types being incorrect.
Would the following be a good alternative? 
```typescript
resolve: async (root, args, context, info) => {
  if (this.options.relayOptions?.nodesQueryOptions?.resolve) {
    return (await this.options.relayOptions?.nodesQueryOptions?.resolve(
      this,
      args,
      context,
      info,
    )) as Promise<ObjectParam<SchemaTypes>>[];
  }

  return (await resolveNodes(this, context, info, args.ids as string[])) as Promise<
    ObjectParam<SchemaTypes>
  >[];
```

hayes replied:
I'd rather just cast the type of the resolve function than incur additional runtime overhead to fix a type issue.   You can probably just do 

resolve:
```ts
  this.options.relayOptions?.nodesQueryOptions?.resolve as never ??
  (async (root, args, context, info) =>
    (await resolveNodes(this, context, info, args.ids as string[])) as Promise<
      ObjectParam<SchemaTypes>
    >[]),
```

It's a bit of a hack, but getting these types right internally is less important than making sure the public API is properly type-safe and ensuring that the runtime overhead is as low as possible

hayes commented on packages/plugin-relay/src/types.ts:
```diff
@@ -49,20 +49,24 @@ export type RelayPluginOptions<Types extends SchemaTypes> = EmptyToOptional<{
     'args' | 'resolve' | 'type'
-  >;
+  > & {
+    resolve?: (ids: string[], context: Types['Context']) => MaybePromise<Types['Objects'][]>;
```
you can remove this, and just remove the `resolve` from the `Omit` above if you take all the normal resolve args

JipSterk replied:
i could remove `resolve` from `Omit` but then the resolve options is not nullable so users would have to define it. If people upgrade the package it would break their code.

hayes replied:
ah right, you probably still want to match the resolve type to be similar to the original resolve method (all 4 args, plus supporting the full variety of supported return types).  You can probably just look at how resolve is implemented on the original options object

hayes commented on packages/plugin-relay/src/types.ts:
```diff
@@ -49,20 +49,24 @@ export type RelayPluginOptions<Types extends SchemaTypes> = EmptyToOptional<{
       FieldNullability<[unknown]>,
       { ids: InputFieldRef<InputShape<Types, 'ID'>[]> },
-      Promise<unknown>[]
+      Promise<Types['Objects']>[]
```
I think `unknown` is the best we can do here.  It's impossible to type check this.  The type you are using there works for object types explicitly defined in builder setup, but won't work for nodes created with any of the `ref` based apis.

hayes commented on packages/plugin-relay/tests/examples/custom-resolve/builder.ts:
```diff
@@ -0,0 +1,59 @@
+    clientMutationIdInputOptions: {
+      description: 'client id input',
+    },
+    mutationInputArgOptions: {
```
probably don't need to duplicate all these options that aren't being tested by this example

hayes commented:
awesome, this looks like a great start!  Thanks for working on this.

I think it's actually pretty close.  

> the only thing i do not know is how to implement type inference.

Can you elaborate on what you mean by this?  I am guessing you mean that type-checking the custom resolver?  Unfortunately I don't think this is possible in a way that is actually useful.

> For now i copied one of the tests & examples but i could provide other tests if needed.

I don't think we need a lot more tests, but simplifying the tests so that there aren't so many unused types/options/fields might be good.  Right now the example used for the test has tons of things that make it hard to figure out what is relevant to that specific test


I think one thing that may be missing (or helpful) is a way to fallback to the default resolver.  I think what you really want is a way to handle just specific ids, and then use the default loaders you already defined for most of the resolvers right?  I am not sure what that API would look like, and it might be something to add later, but I imagine it could be quite useful.

JipSterk commented:
the return type of resolve is now `unknown` and removed a lot of options on the test. Getting the tests to pass however isn't something i yet succeeded in. i've modified the arg to be of the type `globalID` & `globalIDList` instead of `id` and `idList` but it throws an error [here](https://github.com/hayes/pothos/blob/3dac1abea8a08319af47fd6f3afb3d705d098306/packages/core/src/config-store.ts#L268) because it cannot resolve the ref because its `undefined` any idea what could cause this?

hayes commented:
@JipSterk you can merge this: https://github.com/JipSterk/pothos/pull/1 to fix the issues with tests not running, just did a small refactor for how resolveNodes works to avoid extra re-encoding of global ids

hayes reviewed (approved):
This looks like a good first version!  I'll probably try to add in a way to fall back to the original resolve node (so you can resolve just specific types with a custom resolver, and fallback to the default for everything else).  

I think all thats missing from this PR is a changeset
